File: Ahorn/entities/TowerLantern.jl
module TowerLantern

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/TowerLantern" TowerLantern(x::Integer, y::Integer, flickerIntensity::Number=0.1, lanternId::String="", lightColor::String="FFA500", lightRadius::Integer=80, startLit::Bool=false)

const placements = Ahorn.PlacementDict(
    "unlit" => Ahorn.EntityPlacement(TowerLantern),
    "lit" => Ahorn.EntityPlacement(TowerLantern)
)

function Ahorn.selection(entity::TowerLantern)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TowerLantern, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/towerLantern.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local towerLantern = {}

towerLantern.name = "MaggyHelper/TowerLantern"
towerLantern.depth = -50
towerLantern.fieldInformation = {
    lightRadius = {
        minimumValue = 0
    },
    flickerIntensity = {
        minimumValue = 0,
        maximumValue = 1
    },
    lightColor = {
        fieldType = "color"
    }
}
towerLantern.placements = {
    {
        name = "unlit",
        data = {
            lightRadius = 80,
            flickerIntensity = 0.1,
            lanternId = "",
            startLit = false,
            lightColor = "FFA500"
        }
    },
    {
        name = "lit",
        data = {
            lightRadius = 80,
            flickerIntensity = 0.1,
            lanternId = "",
            startLit = true,
            lightColor = "FFA500"
        }
    }
}

function towerLantern.sprite(room, entity)
    local state = entity.startLit and "lit" or "unlit"
    local sprite = drawableSprite.fromTexture("objects/tower_lantern/" .. state, entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    
    if entity.startLit then
        sprite.color = {1.0, 0.65, 0.0, 1.0}
    end
    
    return sprite
end

function towerLantern.selection(room, entity)
    return utils.rectangle(entity.x - 6, entity.y - 20, 12, 20)
end

return towerLantern
